### PR TITLE
Support for minDate and maxDate for months and years calendar.

### PR DESCRIPTION
**Pull request by richard-lopes**
Added:
- support for minDate and maxDate for months calendar.
- support for minDate and maxDate for years calendar.
- added 2 more examples in the basic example.

minDate and maxDate only apply a `disabled` classname to the filtered items in the calendar in the exact same way it is done for the full fledged calendar. It also works for calendar that drill down from years, to months, to days: minDate and maxDate applies across.

As for the original unit tests, there is no tests added for minDate and maxDate (maybe as they only affect the rendered markup ?)

### CODE REVIEW

**Comment by richard-lopes** [comment]
Can someone give it a try ? Take a look ?

**Comment by richard-lopes** [comment]
Ok I couldn't be satisfied without tests so I wrote 2 unit tests anyway.

File: src/DateTimePickerMonths.js [hunk 1]
```diff
@@ -12,23 +12,32 @@ export default class DateTimePickerMonths extends Component {
...
     mode: PropTypes.oneOf([Constants.MODE_DATE, Constants.MODE_MONTH, Constants.MODE_DATETIME])
   }
 
   renderMonths = () => {
-    var classes, i, month, months, monthsShort;
+    var classes, i, month, months, monthsShort, minDate, maxDate, currentMonth;
     const onClick = this.props.mode === Constants.MODE_MONTH ? this.props.setSelectedMonth : this.props.setViewMonth;
     month = this.props.selectedDate.month();
     monthsShort = moment.monthsShort();
+    minDate = this.props.minDate ? this.props.minDate.clone().subtract(1, "months") : this.props.minDate;
```
Comment (camjackson) [inline src/DateTimePickerMonths.js]: Is the subtraction because months are zero-based?

File: src/DateTimePickerMonths.js [hunk 2]
```diff
@@ -12,23 +12,32 @@ export default class DateTimePickerMonths extends Component {
...
+    maxDate = this.props.maxDate ? this.props.maxDate.clone() : this.props.maxDate;
     i = 0;
     months = [];
+    currentMonth = moment([this.props.viewDate.year()]);
```
Comment (camjackson) [inline src/DateTimePickerMonths.js]: Might be clearer to do `moment().year(this.props.viewDate.year())`?

File: src/DateTimePickerMonths.js (hunk 1 above)
Comment (richard-lopes) [inline src/DateTimePickerMonths.js]: Nope. I tried the datetimepicker when drilling down from years, to months, to days. As days substract 1 day so the render enables the picks from the right day, I had to substract 1 month so it starts from the right place too. Same for years I believe. In the examples folder I have an example of a range from -3 to +3 months (or maybe -1 to +1 as I also wrote another example in myob_widgets): with that substract it does enable 3 months before current month and 3 months after current month.

File: src/DateTimePickerMonths.js (hunk 2 above)
Comment (richard-lopes) [inline src/DateTimePickerMonths.js]: Yes true I will change that. Moment() offer several option: the year one is indeed more cryptic.

File: src/DateTimePickerMonths.js [hunk 3]
```diff
@@ -12,23 +12,32 @@ export default class DateTimePickerMonths extends Component {
...
     while (i < 12) {
       classes = {
         month: true,
         "active": i === month && this.props.viewDate.year() === this.props.selectedDate.year()
       };
+      if ((minDate && currentMonth.isBefore(minDate)) || (maxDate && currentMonth.isAfter(maxDate))) {
+        classes.disabled = true;
+      }
```
Comment (camjackson) [inline src/DateTimePickerMonths.js]: You could inline this into the previous statement, i.e.
```js
 classes = {
  month: true,
  "active": i === month && this.props.viewDate.year() === this.props.selectedDate.year()
  disabled: (minDate && currentMonth.isBefore(minDate)) || (maxDate && currentMonth.isAfter(maxDate))
};
```

File: src/DateTimePickerMonths.js [hunk 4]
```diff
@@ -12,23 +12,32 @@ export default class DateTimePickerMonths extends Component {
...
+    var classes, i, month, months, monthsShort, minDate, maxDate, currentMonth;
     const onClick = this.props.mode === Constants.MODE_MONTH ? this.props.setSelectedMonth : this.props.setViewMonth;
     month = this.props.selectedDate.month();
     monthsShort = moment.monthsShort();
+    minDate = this.props.minDate ? this.props.minDate.clone().subtract(1, "months") : this.props.minDate;
+    maxDate = this.props.maxDate ? this.props.maxDate.clone() : this.props.maxDate;
     i = 0;
     months = [];
+    currentMonth = moment([this.props.viewDate.year()]);
     while (i < 12) {
```
Comment (camjackson) [inline src/DateTimePickerMonths.js]: I know it's not your code, but I feel like this loop might be nicer re-written as a for loop.

File: src/DateTimePickerMonths.js (hunk 3 above)
Comment (richard-lopes) [inline src/DateTimePickerMonths.js]: Totally.

File: src/DateTimePickerMonths.js (hunk 4 above)
Comment (richard-lopes) [inline src/DateTimePickerMonths.js]: Well I didn't want to "alter" the original feel of the code but I noticed a few places that could need some refactoring as well. Ok I can do that too.

File: src/DateTimePickerYears.js [hunk 5]
```diff
@@ -22,7 +26,10 @@ export default class DateTimePickerYears extends Component {
         old: i === -1 | i === 10,
         active: this.props.selectedDate.year() === year
       };
-      years.push(<span className={classnames(classes)} key={year}onClick={this.props.setViewYear}>{year}</span>);
+      if ((minDate && year < minDate.year()) || (maxDate && year > maxDate.year())) {
+        classes.disabled = true;
+      }
```
Comment (camjackson) [inline src/DateTimePickerYears.js]: Same here, could inline this.

File: src/DateTimePickerMonths.js (hunk 1 above)
Comment (camjackson) [inline src/DateTimePickerMonths.js]: Ok, makes sense. As long as behaviour for min/max date being inclusive/exclusive is consistent across all 3 levels (day, month, year).

File: src/DateTimePickerMonths.js (hunk 2 above)
Comment (camjackson) [inline src/DateTimePickerMonths.js]: Yeah, I sometimes wish moment.js had fewer ways to do the same thing :)

File: src/DateTimePickerMonths.js (hunk 4 above)
Comment (camjackson) [inline src/DateTimePickerMonths.js]: Not a critical thing for this PR, just something that caught my eye while reading the code. We don't need to refactor the entire repo all at once.

**Comment by camjackson** [comment]
Just ran the thing to make sure it works. Looks good! :+1:

**Comment by richard-lopes** [comment]
:-)

File: src/DateTimePickerYears.js (hunk 5 above)
Comment (richard-lopes) [inline src/DateTimePickerYears.js]: Will do.

File: src/DateTimePickerMonths.js (hunk 4 above)
Comment (richard-lopes) [inline src/DateTimePickerMonths.js]: I will do all those changes so we can merge this. Thanks for taking the time to look at this. Yes we can refactor more later.

File: src/DateTimePickerMonths.js [hunk 6]
```diff
@@ -12,23 +12,28 @@ export default class DateTimePickerMonths extends Component {
...
+    var classes, month, months, monthsShort, minDate, maxDate, currentMonth;
     const onClick = this.props.mode === Constants.MODE_MONTH ? this.props.setSelectedMonth : this.props.setViewMonth;
     month = this.props.selectedDate.month();
     monthsShort = moment.monthsShort();
-    i = 0;
+    minDate = this.props.minDate ? this.props.minDate.clone() : this.props.minDate;
+    maxDate = this.props.maxDate ? this.props.maxDate.clone().add(1, "months") : this.props.maxDate;
     months = [];
-    while (i < 12) {
+    currentMonth = moment([this.props.viewDate.year(), 1, 1]);
```
Comment (richard-lopes) [inline src/DateTimePickerMonths.js]:
```Javascript
currentMonth = moment([this.props.viewDate.year(), 1, 1]);
```
Using `moment().year(...)` had unexpected consequences as the original idea was to set the date to the first day of the first month of the year. Calling `year(...)` would keep current date and just change the year.

**Comment by richard-lopes** [comment]
@camjackson Ran my visual tests (`npm run examples`) and the unit tests again. Should be good.

File: src/DateTimePickerMonths.js (hunk 6 above)
Comment (camjackson) [inline src/DateTimePickerMonths.js]: Oh right, of course.

Comment (camjackson) [inline src/DateTimePickerMonths.js]: Should the month number be 1 or 0 though?

Comment (richard-lopes) [inline src/DateTimePickerMonths.js]: Yes you are right. Good catch.
I  am back home. Will fix and retest that tomorrow morning.

On 14 Dec 2015 5:09 pm, Cam Jackson <notifications@github.com> wrote:

In src/DateTimePickerMonths.js<https://github.com/MYOB-Technology/react-bootstrap-datetimepicker/pull/3#discussion_r47460208>:

>      months = [];
> -    while (i < 12) {
> +    currentMonth = moment([this.props.viewDate.year(), 1, 1]);


Should the month number be 1 or 0 though?

—
Reply to this email directly or view it on GitHub<https://github.com/MYOB-Technology/react-bootstrap-datetimepicker/pull/3/files#r47460208>.

________________________________

MYOB values your privacy. For information on how we collect and manage your personal information please visit our privacy pages for Australia<www.myob.com.au/privacy-disclosure> and New Zealand<www.myob.co.nz/privacy>. If you are an MYOB client and wish to update your communication preferences, or unsubscribe from communications, please visit my.MYOB.com. This email and any attachment are confidential. If you are not the intended recipient, please notify MYOB by reply email and delete this email. Please note that you must not access or use this email or any information in it. MYOB accepts no liability for viruses in this email or in any attachment to it.

**Comment by richard-lopes** [comment]
@camjackson Ready for one more look

**Pull request merged**